### PR TITLE
feat: CURP WAL Pipeline

### PR DESCRIPTION
**Pull request by bsbds**
Based on : #563
Please briefly answer these questions:

* what problem are you trying to solve? (or if there's no problem, what's the motivation for this change?)

  Implement CURP WAL

* what changes does this pull request make?

  This PR implements the WAL pipeline part, the file pipeline is used for pre-allocating files for WAL segment.

* are there any non-obvious implications of these changes? (does it break compatibility with previous versions, etc)

### CODE REVIEW

File: curp/src/server/storage/wal/pipeline.rs [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+
+        let (file_tx, file_rx) = flume::bounded(1);
+        let stop_event = Event::new();
+        let mut stop_listener = stop_event.listen();
+        let dir_c = dir.clone();
+
+        let handle = tokio::spawn(async move {
+            let mut file_count = 0;
+            loop {
+                let file = Self::alloc(&dir_c, file_size, &mut file_count)?;
```
Comment (themanforfree) [inline curp/src/server/storage/wal/pipeline.rs:49]: I think it's better to pass `usize` directly here and then increment `file_count` inside the loop

File: curp/src/server/storage/wal/pipeline.rs [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+                        // The receiver is already dropped, stop this task
+                        if let Err(e) = result {
+                            break;
+                        }
+                    }
+                }
+            }
+            Self::clean_up(&dir_c)?;
+            Ok(())
+        });
```
Comment (themanforfree) [inline curp/src/server/storage/wal/pipeline.rs:64]: this task is cancel safe, we can remove `stop_listener` and direct abort it

File: curp/src/server/storage/wal/pipeline.rs (hunk 1 above)
Comment (bsbds) [reply curp/src/server/storage/wal/pipeline.rs]: done

File: curp/src/server/storage/wal/pipeline.rs (hunk 2 above)
Comment (bsbds) [reply curp/src/server/storage/wal/pipeline.rs]: done

File: curp/src/server/storage/wal/pipeline.rs [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+                    }
+                    Ok(Ok(_)) => return Poll::Ready(None),
+                }
+            }
+            return Poll::Ready(None);
+        }
+
+        if self.stopped {
+            return Poll::Ready(None);
+        }
```
Comment (Phoenix500526) [inline curp/src/server/storage/wal/pipeline.rs:128]: We can merge this piece of code. The `drop` method will invoke the `stop` method, aborting the handle and setting the stop flag to true. Therefore, the stop flag must be true when the `is_finished` returns true. If the `is_finished` returns false, the stop flag is still true. We can bank on this stop flag to cancel the stream.

File: curp/src/server/storage/wal/pipeline.rs [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+            dir,
+            file_size,
+            file_stream: file_rx.into_stream(),
+            stopped,
+        })
+    }
+
+    /// Stops the pipeline
+    pub(super) fn stop(&mut self) {
+        self.stopped.store(true, Ordering::SeqCst);
```
[the remark this answers is not in the document]
Comment (bsbds) [reply curp/src/server/storage/wal/pipeline.rs]: It seems `Ordering::Relaxed` is sufficient in this case, fixed.

**Comment by Phoenix500526** [comment]
Please add a signed-off message in these commits

File: curp/src/server/storage/wal/pipeline.rs [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+        let (file_tx, file_rx) = flume::bounded(1);
+        let stop_event = Event::new();
+        let dir_c = dir.clone();
+        let stopped = Arc::new(AtomicBool::new(false));
+        let stopped_c = Arc::clone(&stopped);
+
+        let _ignore = tokio::spawn(async move {
+            let mut file_count = 0;
+            loop {
+                let file = match Self::alloc(&dir_c, file_size, file_count) {
```
Comment (iGxnon) [inline curp/src/server/storage/wal/pipeline.rs:52]: `Self::alloc` seems to be slow (preallocation) and it is blocking (i.e. no `.await`). Could we convert it into async io or spawn this task in blocking threads pool(`tokio::spawn_blocking`)?

Comment (iGxnon) [reply curp/src/server/storage/wal/pipeline.rs:52]: This may block the Tokio scheduler every time the log is sealed and generate the next log file.

Comment (bsbds) [reply curp/src/server/storage/wal/pipeline.rs]: Tokio file does not seems to support preallocation, I refactored this to run in another thread.

File: crates/curp/src/server/storage/wal/pipeline.rs [hunk 6]
```diff
@@ -0,0 +1,183 @@
...
+use tokio_stream::Stream;
+use tracing::error;
+
+use super::util::LockedFile;
+
+/// The temp file extension
+const TEMP_FILE_EXT: &str = ".tmp";
+
+/// The file pipeline, used for pipelining the creation of temp file
+pub(super) struct FilePipeline {
```
Comment (Phoenix500526) [inline crates/curp/src/server/storage/wal/pipeline.rs:26]: Can a user create more than one `FilePipeline`? If so, what will happen?

File: crates/curp/src/server/storage/wal/pipeline.rs [hunk 7]
```diff
@@ -0,0 +1,183 @@
...
+    stopped: Arc<AtomicBool>,
+}
+
+impl FilePipeline {
+    /// Creates a new `FilePipeline`
+    pub(super) fn new(dir: PathBuf, file_size: u64) -> io::Result<Self> {
+        Self::clean_up(&dir)?;
+
+        let (file_tx, file_rx) = flume::bounded(1);
+        let stop_event = Event::new();
```
Comment (Phoenix500526) [inline crates/curp/src/server/storage/wal/pipeline.rs:43]: What's this `stop_event` for?

File: crates/curp/src/server/storage/wal/pipeline.rs (hunk 6 above)
Comment (bsbds) [reply crates/curp/src/server/storage/wal/pipeline.rs:26]: The user don't create a pipeline, it only act as a WAL componet and will only be created once.

File: crates/curp/src/server/storage/wal/pipeline.rs (hunk 7 above)
Comment (bsbds) [reply crates/curp/src/server/storage/wal/pipeline.rs]: it serve no use, removed

**Review by Phoenix500526 — APPROVED** [review]
LGTM

**Review by iGxnon — APPROVED** [review]

**Pull request merged**